probably/pgcl/parser.py

```diff
@@ -105,6 +130,22 @@ def _doc_parser_grammar():
 _doc_parser_grammar.__doc__ = "The Lark grammar for pGCL::\n" + _PGCL_GRAMMAR + "\n\nThis function only exists for documentation purposes and should never be called in code."
 
 
+# Collect parameter information here.
+parameters: Dict[Var, Type] = dict()
```

Comment:
This global state should not be public and be prefixed with `_`. I also do not understand why it even exists. Checking whether parameters are assigned to should be done in the type checker, not done as part of the parser. 

This global state also leads to a bug because it is never cleared: After parsing one program, a second program will have the old parameters in this dict and so is wrongly restricted in variable assignments.

Reply:
Removing this dict completely would mean that the `Program` that the parser produces no longer contains the information which `VarExpr`s are parameters. This isn't necessarily a problem, as the `Program` still contains the names of all parameters, but it would mean we have to rewrite some unrelated code as well.

For example, completely removing `_parameters` could look like this (this also removes some checks the parser does, but they belong in the typechecker anyways): 9943cb9eaf7519affb2c8de683604ad1923d4a01